### DISCUSSION TITLE
Support reading array of multiple sub protocols from client

### DISCUSSION THREAD

**Discussion by dushaniw** [enisdenjo/graphql-ws#242] [Q&A] [answered]
### Story

I am implementing a middle-ware gateway component to manage GraphQL API calls exchanged between client and server. I am trying make my gateway to work with the latest "graphql-ws" protocol, as well as the legacy "subscriptions-transport-ws" protocol (As there will be old users who use it still) I would like to know in default case, if its possible for the server side to accept an array of sub protocols and choose whatever it supports and send that back to client. The [MDN explanation](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#subprotocols) for web-sockets states that

```
client has to ask for a specific subprotocol. To do so, it will send something like this as part of the original handshake:

Sec-WebSocket-Protocol: soap, wamp

or, equivalently:

Sec-WebSocket-Protocol: soap
Sec-WebSocket-Protocol: wamp

Now the server must pick one of the protocols that the client suggested and it supports. If there is more than one, send the first one the client sent. 
``` 
My sample server code looks like this (same as your sample server code)
```
const { WebSocket } = require('ws');
const { buildSchema } = require ('graphql');
const { useServer } = require ('graphql-ws/lib/use/ws');

// Construct a schema, using GraphQL schema language
const schema = buildSchema(`
  type Query {
    hello: String
  }
  type Subscription {
    greetings: String
  }
`);

// The roots provide resolvers for each GraphQL operation
const roots = {
  query: {
    hello: () => 'Hello World!',
  },
  subscription: {
    greetings: async function* sayHiIn5Languages() {
      for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
        yield { greetings: hi };
      }
    },
  },
};

const server = new WebSocket.Server({ port: 40000, path : '/graphql', });

  useServer(
    // from the previous step
    { schema, roots },
    server,
  );
  
console.log('Listening to port 4000');
```
If I write a simple nodejs server using "graphql-ws" and "ws", and if my client send the subprotocol header values as "graphql-transport-ws, graphql-ws" server code always read the first protocol and works fine, but if the header values are in order "graphql-ws, graphql-transport-ws" it doesn't work because the first unsupported protocol is picked. 

As I checked, "graphql-ws" also support writing in this in a customizable way (support legacy subscriptions-transport-ws). But, server side developer should explicitly write the protocol reading part for that. Shouldn't choosing values from protocol header array be the default behavior even for simple cases?

### Acceptance criteria
Client will be able to send comma separated multiple sub protocol header values in any order.
In simple case, graphql-ws libraray server code will be able to read and pick one of the protocols that the client suggested and it supports (Eg: "graphql-ws").

**Comment by enisdenjo** [comment, marked as answer]
Hi there,

the client is fixed using `"graphql-transport-ws"` subprotocol intentionally because it does not know which subprotocol the server chose when multiple are supplied. The client just knows if the connection was accepted or not. This means that sending both subprotocols (`["graphql-transport-ws", "graphql-ws"]`) from the `createClient` will break if the latter is selected by the server due to subprotocol mismatch.

**Instead, what you should do is check one by one.** To check if this library ("graphql-transport-ws") is supported, you can do something like:
```ts
function supportsGraphQLTransportWS(url: string): Promise<boolean> {
  return new Promise((resolve) => {
    const client = createClient({
      url, // enpoint you want to check
      retryAttempts: 0, // dont retry, fail immediately
      lazy: false, // connect as soon as the client is created, so as soon as this function is called
      on: {
        closed: () => resolve(false), // connection rejected, NOT supported
        connected: () => {
          resolve(true); // connected = supported
          client.dispose(); // dispose of client after check
        },
      },
    });
  });
}

const supported = await supportsGraphQLTransportWS('wss://somwhere.out/there');
if (supported) {
  // use graphql-ws
} else {
  // use subscriptions-transport-ws
}
```

Furthermore, regarding the server-side, you can indeed support both the legacy and the current subprotocol. Please refer to the ["ws server usage with subscriptions-transport-ws backwards compatibility" recipe](https://github.com/enisdenjo/graphql-ws#ws-backwards-compat).

**Comment by enisdenjo** [comment]
Added a  ["Client usage supported check" recipe](https://github.com/enisdenjo/graphql-ws#supported-check) with https://github.com/enisdenjo/graphql-ws/commit/c47dcf06fa5d56bf4a8b83d33c2b82f1231f67ef.

**Comment by leszekhanusz** [comment]
Hi,

I'm the maintainer of [graphql-python/gql](https://github.com/graphql-python/gql) and I really think you should reconsider and allow graphql-ws to work if its subprotocol is present later in the subprotocols provided list by the client.

The client can and should read the answer of the backend in the `Sec-WebSocket-Protocol` http response header to know which protocol was chosen by the backend.

It should work correctly right now with the current gql code.

If you allow this, it would permit the client to support both the apollo and graphql-ws protocol automatically (currently it works only with your backwards-compatible recipe)

**Reply by enisdenjo** [reply]
~~I guess you didn't read [my answer](https://github.com/enisdenjo/graphql-ws/discussions/242#discussioncomment-1465054) in detail:~~

> ~~...the client is fixed using `"graphql-transport-ws"` subprotocol intentionally because it does not know which subprotocol the server chose when multiple are supplied. The client just knows if the connection was accepted or not. This means that sending both subprotocols (`["graphql-transport-ws", "graphql-ws"]`) from the `createClient` will break if the latter is selected by the server due to subprotocol mismatch.~~

~~In essence, there's nothing to _reconsider_, it is simply not possible given the current state of things - the browser does not communicate which protocol is accepted when supplying multiple on the connection upgrade request.~~

Seems like I was ill-informed, the [WebSocket instance indeed has a `protocol` property](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/protocol). I'll investigate further and report back!

**Reply by enisdenjo** [reply]
My question for you, @leszekhanusz, what should `graphql-ws` client do if an unsupported protocol is selected? How would you even perform that check using the `graphql-ws` client and _why_ would you even use `graphql-ws` for that check?

**Reply by leszekhanusz** [reply]
The client sends the supported subprotocols to the backend. By default, graphql-python/gql send both apollo and graphql-ws subprotocols.

the graphql-ws backend should only check if its own subprotocol is present in the list and if yes, then allow the connection, it should not care whether there are other things in the list.

the graphql-ws backend will send back to the client always its `graphql-transport-ws` protocol in the `Sec-WebSocket-Protocol` response header. It is currently the case with the current graphql-ws code.

In the client, [we save the protocol selected by the backend](https://github.com/graphql-python/gql/blob/12fc895bd45c297ce7664358fcdebb216fe11cac/gql/transport/websockets.py#L475) and depending on this answer from the backend, [we will parse the answer from the backend differently](https://github.com/graphql-python/gql/blob/12fc895bd45c297ce7664358fcdebb216fe11cac/gql/transport/websockets.py#L413).

If we don't receive a valid response from the server, then we fall back to the apollo code.

**Reply by enisdenjo** [reply]
Ah ok, so this is backend focused - you're absolutely right, multiple subprotocols should be supported! PR incoming...

**Reply by enisdenjo** [reply]
Server now supports and handles multiple subprotocols correctly. @leszekhanusz try it out in version [v5.6.1](https://github.com/enisdenjo/graphql-ws/releases/tag/v5.6.1)! 😄

**Reply by leszekhanusz** [reply]
That's great! It works well :rocket: Thank you!

**Reply by leszekhanusz** [reply]
After further testing, I noticed a very minor issue. The error message printed on the console if the subprotocol is not found is:

```
WebSocket protocol error occured. It was most likely caused due to an unsupported subprotocol "" requested by the client. graphql-ws implements exclusively the "graphql-transport-ws" subprotocol, please make sure that the client implements it too.
```

In your code you use `$socket.protocol` which is empty now:
```
console.warn(                                                                                                   
  `WebSocket protocol error occured. It was most likely caused due to an ` +                                    
    `unsupported subprotocol "${socket.protocol}" requested by the client. ` +                                  
    `graphql-ws implements exclusively the "${GRAPHQL_TRANSPORT_WS_PROTOCOL}" subprotocol, ` +                  
    'please make sure that the client implements it too.',                                                      
); 
```

**Reply by enisdenjo** [reply]
That's expected. As per the specification:
> /subprotocol/
          Either a single value representing the subprotocol the server
          is ready to use or null.  The value chosen MUST be derived
          from the client's handshake, specifically by selecting one of
          the values from the |Sec-WebSocket-Protocol| field that the
          server is willing to use for this connection (if any).  If the
          client's handshake did not contain such a header field or if
          the server does not agree to any of the client's requested
          subprotocols, the only acceptable value is null.  The absence
          of such a field is equivalent to the null value (meaning that
          if the server does not wish to agree to one of the suggested
          subprotocols, it MUST NOT send back a |Sec-WebSocket-Protocol|
          header field in its response).  The empty string is not the
          same as the null value for these purposes and is not a legal
          value for this field.  The ABNF for the value of this header
          field is (token), where the definitions of constructs and
          rules are as given in RFC2616.

_[RFC 6455: 4.2.2. Sending the Server's Opening Handshake](https://www.rfc-editor.org/rfc/rfc6455#section-4.2.2)_

A proper response from the server, when presented with unsupported subprotocol(s) from the client, is to omit the `Sec-WebSocket-Protocol` or set its value to `null`, proceed with the upgrade and have the client handle the lack of an agreed subprotocol. It's then the client's responsibility to act accordingly.

For example, [`ws` immediately aborts the handshake with a `"Server sent no subprotocol"` error](https://github.com/websockets/ws/blob/c9d5436500fad16493a2cc62a0ce6daed83c9129/lib/websocket.js#L890). However, if the client still decides to proceed with the upgrade - the `socket.protocol` on the server will be empty.

I suggest your client does the same, abort the handshake if no subprotocol is agreed upon.
